client/src/components/DisplayTable/DisplayTableUtils.js

```diff
@@ -62,3 +63,204 @@ export const DisplayTableDownloadCsv = ({ csv_string, table_name }) => (
     )}
   </Fragment>
 );
+
+export const SelectPage = ({
+  current_page,
+  num_pages,
+  change_page,
+  num_col,
+}) => {
+  const get_shifted_option_range = (
+    page_count,
+    current_page,
+    option_window
+  ) => {
+    const window_start = current_page - option_window;
+    const window_end = current_page + option_window;
+    if (window_start < 1) {
+      return _.range(1, window_end + 1 - (window_start - 1));
+    } else if (window_end > page_count) {
+      return _.range(window_start - (window_end - page_count), page_count + 1);
+    } else {
+      return _.range(window_start, window_end + 1);
+    }
+  };
+  const extend_options_left = (option_range, page_count, current_page) => {
+    if (_.first(option_range) === 1) {
+      return [
+        ...option_range,
+        _.last(option_range) + 1,
+        _.last(option_range) + 2,
+      ];
+    } else if (_.first(option_range) === 2) {
+      return [1, ...option_range, _.last(option_range) + 1];
+    } else if (_.first(option_range) === 3) {
+      return [1, 2, ...option_range];
+    } else if (current_page === page_count - 3) {
+      return [
+        1,
+        _.first(option_range) - 2,
+        _.first(option_range) - 1,
+        ...option_range,
+      ];
+    } else if (current_page > page_count - 3) {
+      return [
+        1,
+        _.first(option_range) - 3,
+        _.first(option_range) - 2,
+        _.first(option_range) - 1,
+        ...option_range,
+      ];
+    } else {
+      return option_range;
+    }
+  };
+  const extend_options_right = (option_range, page_count, option_window) => {
+    if (
+      _.includes(option_range, page_count - option_window) &&
+      _.last(option_range) !== page_count
+    ) {
+      return [
+        ...option_range,
+        ..._.range(_.last(option_range) + 1, page_count + 1),
+      ];
+    } else {
+      return option_range;
+    }
+  };
+  const get_page_options = (page_count, current_page, option_window = 2) => {
+    const raw_page_options = _.chain(
+      //without ellipsis
```

Comment:
I think due to the nature of this, it's really hard to follow the code for all the different cases. Would you mind adding very short comments explaining what that line is doing, for some of the code? Like you did here `without ellipsis`